torch/_meta_registrations.py

```diff
@@ -143,3 +132,86 @@ def meta_repeat_interleave_Tensor(repeats, output_size=None):
             "cannot repeat_interleave a meta tensor without output_size"
         )
     return repeats.new_empty(output_size)
+
+@torch.library.impl(meta_lib, "index.Tensor")
+def meta_index_Tensor(self, indices):
+    # aten::index is the internal advanced indexing implementation
+    check(len(indices) <= self.ndim, lambda: f"too many indices for tensor of dimension {self.ndim} (got {len(indices)})")
+    # checkIndexTensorTypes
+    result = []
+    for i, index in enumerate(indices):
+        if index is not None:
+            check(
+                index.dtype in [torch.long, torch.int8, torch.bool],
+                lambda: "tensors used as indices must be long, byte or bool tensors"
+            )
+            if index.dtype in [torch.int8, torch.bool]:
+                nonzero = index.nonzero()
+                for j in range(index.ndim):
+                    result.append(nonzero.select(1, j))
```

Comment:
Missing a few checks in the boolean case as the dimensions of the mask need to agree with the dimensions of the tensor.

I think the correct way of doing this is as follows modulo errata I have had (certainly not tested)
```suggestion
                k = len(result)
                # TODO. This should be an IndexError. Consider extending check with an
                # exception_type=RuntimeError parameter
                check(k + index.ndim <= self.ndim,
                      lambda: f"too many indices for tensor of dimension {self.ndim}")
                for j in range(index.ndim):
                    # IndexError as well
                    check(index[j] <= self.shape[k + j],
                          lambda: f"The shape of the mask {index.shape} at index {i} does not match the shape of the indexed tensor {self.shape} at index {k + j}")
                    result.append(nonzero.select(1, j))
```